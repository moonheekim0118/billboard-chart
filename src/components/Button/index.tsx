import React from 'react';
import styled from 'styled-components';

interface Props {
    onClick: (e: any) => void;
    title: string;
    disabled: boolean;
    type: string;
}

const Button = (props: Props) => {
    return (
        <Container
            type={props.type}
            disabled={props.disabled}
            onClick={props.onClick}>
            {props.title}
        </Container>
    );
};

const Container = styled.button`
    background-color: #ecc422;
    color: #fff;
    font-size: 1.1em;
    font-weight: bold;
    border: none;
    border-top-right-radius: 10px;
    border-bottom-right-radius: 10px;
    padding: 10px 15px;

    cursor: pointer;
    transition: background-color 0.5s ease;

    &:disabled {
        background-color: #f3d972;
        cursor: not-allowed;
    }

    &:hover {
        background-color: #d3ad12;
    }
`;

export default Button;
