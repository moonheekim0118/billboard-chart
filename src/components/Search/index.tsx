import React, { useCallback } from 'react';
import { setItem } from '../../util/localStorage';
import Button from '../Button';
import SearchHistory from '../SearchHistory';
import useToggle from '../../hooks/useToggle';
import useInput from '../../hooks/useInput';
import styled from 'styled-components';

// TODO : sessionStorage에서 keyword 가져와서 input initlaValue로 넣기

interface Props {
    fetchAPI: (value: string) => void;
    keyword: string;
}
const Search = (props: Props) => {
    const [value, onInput, validation] = useInput(props.keyword);
    const [showHistory, setShowHistory] = useToggle();

    const onSubmit = useCallback(
        async (e) => {
            e.preventDefault();
            setItem('searchHistory', value);
            await props.fetchAPI(value);
        },
        [value]
    );

    return (
        <Form>
            <Input value={value} onChange={onInput} />
            <Button
                title={'submit'}
                onClick={onSubmit}
                type={'submit'}
                disabled={validation}
            />
            <HistoryContainer>
                <SearchHistory />
            </HistoryContainer>
        </Form>
    );
};

const Form = styled.form`
    text-align: center;
    display: flex;
    position: relative;
    padding: 20px 0px;
    width: 100%;
`;

const HistoryContainer = styled.div.attrs({ tabindex: '0' })`
    display: none;
    position: absolute;
    bottom: -80px;
    width: 100%;

    &:hover {
        display: flex;
    }

    &:focus-within {
        display: flex;
    }
`;

const Input = styled.input.attrs({ type: 'text' })`
    width: 100%;
    padding: 10px 15px;
    font-size: 1.1em;
    border: none;
    border-radius: 10px;

    &:focus {
        outline: none;
    }

    &:focus ~ ${HistoryContainer} {
        display: flex;
    }
`;

export default Search;
