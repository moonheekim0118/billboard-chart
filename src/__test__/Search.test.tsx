import 'jsdom-global/register';
import React from 'react';
import Button from '../components/Button';
import Search from '../components/Search';
import { mount } from 'enzyme';

describe('<Search/>', () => {
    let container = null;
    let mockFn = jest.fn();

    it('renders correctly', () => {
        container = mount(<Search fetchAPI={mockFn} keyword={''} />);
    });

    it('matches snapshot', () => {
        expect(container.html()).toMatchSnapshot();
    });

    it('Button should be disabled', () => {
        const submitButton = container.find(Button).first();
        expect(submitButton.prop('disabled')).toBeTruthy();
    });
});
