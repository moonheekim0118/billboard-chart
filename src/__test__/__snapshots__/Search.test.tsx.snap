// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search/> matches snapshot 1`] = `"<form class=\\"sc-jSgupP jIA-deX\\"><input type=\\"text\\" class=\\"sc-iBPRYJ fFPSQN\\" value=\\"\\"><div class=\\"sc-gKsewC hqwfZo\\"><div class=\\"sc-gsTCUz cZSYBp\\"></div></div><div class=\\"sc-fubCfw jggMGH\\"><button type=\\"submit\\" disabled=\\"\\" class=\\"sc-bdfBwQ cLmSPl\\">submit</button></div></form>"`;
